gluten-iceberg/src/main/scala/org/apache/iceberg/spark/source/GlutenIcebergSourceUtil.scala

```diff
@@ -94,15 +97,24 @@ object GlutenIcebergSourceUtil {
       throw new GlutenNotSupportException("Only support iceberg SparkBatchQueryScan.")
   }
 
-  def getPartitionSchema(sparkScan: Scan): StructType = sparkScan match {
+  def getReadPartitionSchema(sparkScan: Scan): StructType = sparkScan match {
     case scan: SparkBatchQueryScan =>
       val tasks = scan.tasks().asScala
       asFileScanTask(tasks.toList).foreach {
         task =>
           val spec = task.spec()
           if (spec.isPartitioned) {
             var partitionSchema = new StructType()
-            val partitionFields = spec.partitionType().fields().asScala
+            val readFields = scan.readSchema().fields.map(_.name).toSet
+            // Iceberg will generate some non-table fields as partition fields, such as x_bucket,
+            // which will not appear in readFields, they also cannot be filtered.
+            val tableFields = spec.schema().columns().asScala.map(_.name()).toSet
+            val partitionFields =
+              spec
+                .partitionType()
+                .fields()
+                .asScala
+                .filter(f => !tableFields.contains(f.name) || readFields.contains(f.name()))
```

Comment:
Can you explain the filter here?  should it be `!tableFields.contains(f.name) && readFields.contains(f.name())`?

Reply:
For iceberg's bucket join, 
spec.partitionType() will return id_bucket, p
tableFields are id, name, p
readFields are id ,name, p
will need both keep id_bucket and p

Reply:
see. Thanks!